src/factories/nodes.ts

```diff
@@ -64,18 +65,20 @@ export async function nodesContainerFactory() {
     highlightSelectedNode()
   })
 
-  async function render(data: RunGraphData): Promise<void> {
+  async function render(data: DataBundle): Promise<void> {
     startWorker()
 
-    runData = data
+    dataBundle = data
     nodesLayout = null
 
     await Promise.all([
-      createNodes(data),
-      createEdges(data),
+      createNodes(),
+      createEdges(),
     ])
 
-    getLayout(data)
+    await createArtifacts()
```

Comment:
Why are we rendering artifacts here in nodes? I expected to see each nodes artifacts passed to it and then each node can be responsible for rendering its own artifacts. That way it can render things relative to itself. It looks like you're already calling `node.renderArtifacts`. So I think it would make sense to just pass the artifacts into the node when rendering the node.

Reply:
Yeah.. I think you're right on adding artifacts to the node's data itself.

The reason for calling the render method from this level is just that the individual node wouldn't know when artifacts have been added to the artifacts array, since they're only subscribed to their individual node data. I thought about some separate data polling service that nodes can subscribe to directly, that feels wack though because then you have a lot of nodes subscribing to an updated event and seeing if they have any.

Reply:
Yeah I think adding artifacts to the date itself is the way to go here if that's not to bad from an api perspective. Otherwise I think that the node render function can just accept the artifacts as well.